### ISSUE TITLE
Column altering in migration from TEXT to LONGTEXT may not apply.

### ISSUE THREAD

**Issue by janokary** [doctrine/dbal#2566]
Based on this
https://github.com/laravel/framework/issues/12363
and my experience here
https://laracasts.com/discuss/channels/laravel/alter-table-change-field-to-longtext

I think there is a problem when migrating a table column from TEXT to LONG TEXT

**Comment by Ocramius** [comment]
@janokary what RDBMs does this apply to? Maybe this is missing schema introspection?

**Comment by janokary** [comment]
It is allways mysql

> version: 5.5.52-0+deb8u1

or

> version: 5.7.11-log

or
>version: 5.7.12-0ubuntu1.1

**Comment by deeky666** [comment]
@janokary I need more information. What is the Doctrine DBAL type before and what should it be after? Also `length` property is is significant.

**Comment by ariews** [comment]
Hi all, I'm facing same issue, using Laravel 5.4, doctrine/dbal v2.5.12, php 5.6.30, mysql  Ver 14.14 Distrib 5.6.33.

before: `$table->text('body')` alter using: `$table->longText('body')->change()`. 

The column type should be `longtext`, but the column type is unchanged.

**Comment by Ocramius** [comment]
@ariews can this be reproduced with DBAL-only?

**Comment by frandieguez** [comment]
Any news on getting this integrated?

**Comment by Ocramius** [comment]
@frandiguez this needs a test case first.

**Comment by Ocramius** [comment]
Can't find a PR for that: can you open or dig up one?

**Comment by cbotsikas** [comment]
@Ocramius here's your PR.

**Comment by cbotsikas** [comment]
I came across this issue trying to change a column in a Laravel application from text to longtext. I traced the migration and the issue was that the change was ignored since [$changedProperties was empty](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/Comparator.php#L235) although the _length was different.
The creation migration in Laravel was like this:
```
Schema::create('myTable', function (Blueprint $table) {
    $table->increments('id');
    $table->text('myColumn')->nullable();
});
```
and the change migration like this:
```
Schema::table('myTable', function (Blueprint $table) {
    $table->longText('myColumn')->change();
});
```

**Comment by cbotsikas** [comment]
I have added some comments [on my PR](https://github.com/doctrine/dbal/pull/3182). There's an issue with a failing test which i feel is not well defined. 
Please have a look and let me know if i'm missing something..

**Comment by uphlewis** [comment]
Some helpful fellow has suggested this workaround, for those who come after me:

https://github.com/laravel/framework/issues/21847#issuecomment-454728171

tldr; change to string column with length > 16777216, since it exceeds varchar max length the column gets converted to longtext. Hacky, but it works.

**Comment by kl83** [comment]
I found another workaround. Just change the column comment, for example:

`$table->mediumText('myColumn')->comment(' ')->change(); // up`
`$table->text('myColumn')->comment('')->change(); // down`

Works perfectly on Ubuntu 18.04, PHP 7.2, MySQL 8, Laravel 5.8

**Comment by ZsgsDesign** [comment]
> I found another workaround. Just change the column comment, for example:
> 
> `$table->mediumText('myColumn')->comment(' ')->change(); // up`
> `$table->text('myColumn')->comment('')->change(); // down`
> 
> Works perfectly on Ubuntu 18.04, PHP 7.2, MySQL 8, Laravel 5.8

I am using the same method to solve the problem. That works.

**Comment by lwitzel** [comment]
Ran across this after trying to diagnose the problem for the last hour. Both hacks work. Fixing $table->longText()->change() would be even better.

**Comment by jm-vasquez** [comment]
> I found another workaround. Just change the column comment, for example:
> 
> `$table->mediumText('myColumn')->comment(' ')->change(); // up`
> `$table->text('myColumn')->comment('')->change(); // down`
> 
> Works perfectly on Ubuntu 18.04, PHP 7.2, MySQL 8, Laravel 5.8

this works as well on ubuntu 16.04, php 7.0 and mariadb 10.2.8, Laravel 5.4

**Comment by stealthhex** [comment]
As of today.. this bug is still in laravel.. 

> I found another workaround. Just change the column comment, for example:
> 
> `$table->mediumText('myColumn')->comment(' ')->change(); // up`
> `$table->text('myColumn')->comment('')->change(); // down`
> 
> Works perfectly on Ubuntu 18.04, PHP 7.2, MySQL 8, Laravel 5.8

As of today.. this bug is still in laravel and this is the fix..

**Comment by holymp2006** [comment]
> I found another workaround. Just change the column comment, for example:
> 
> `$table->mediumText('myColumn')->comment(' ')->change(); // up`
> `$table->text('myColumn')->comment('')->change(); // down`
> 
> Works perfectly on Ubuntu 18.04, PHP 7.2, MySQL 8, Laravel 5.8

As of today, when changing between text types (text, mediumText, or longText), this is still the fix.

**Comment by morozov** [comment]
Closing as stale and missing tests. This will be likely resolved by https://github.com/doctrine/dbal/pull/4659.

**Comment by morozov** [comment]
This bug will be covered by the `Doctrine\DBAL\Tests\Functional\Schema\MySQL::testLobLengthIncrementOverLimit` test in https://github.com/doctrine/dbal/pull/4746.

Additionally,
```php
$column = new Column('myColumn', new TextType(), [
    'length' => 2**15, // TEXT
]);
$desiredTable = new Table('test', [$column]);

$sm = $connection->createSchemaManager();
$sm->dropAndCreateTable($desiredTable);
echo $connection->fetchNumeric('SHOW CREATE TABLE test')[1], PHP_EOL;

// CREATE TABLE `test` (
//   `myColumn` text COLLATE utf8_unicode_ci NOT NULL
// ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci

$column->setLength(2**24); // LONGTEXT

$actualTable = $sm->listTableDetails('test');

$comparator = $sm->createComparator();
$diff = $comparator->diffTable($actualTable, $desiredTable);
$sm->alterTable($diff);
echo $connection->fetchNumeric('SHOW CREATE TABLE test')[1], PHP_EOL;

// CREATE TABLE `test` (
//   `myColumn` longtext COLLATE utf8_unicode_ci NOT NULL
// ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci
```
